// thresh-lib/src/core/Widget.ts
import threshApp from '../..'
import ChildrenRule from './ChildrenRule'
import scheduleUpdate from './scheduleUpdate'
import VNode from './VNode'
import Util from '../shared/Util'
import { RefCallback, ThreshWidget } from '../types/widget'

/**
 * 校验object是否合法
 * @param {any} object 
 */
function checkObjectValid (object: any, noFunc: boolean = false): boolean {
  if (object === null || object === undefined) return true
  if (!Util.isObject(object)) {
    threshApp.onError(new Error('props or state must be an object type'))
    return false
  }
  if (noFunc) {
    for (const key in object) {
      if (Util.isFunc(object[key])) {
        threshApp.onError(new Error(`prop ${key} must not be a function type`))
        return false
      }
    }
  }
  return true
}

/**
 * Widget
 */
// interface WidgetProps {
//   key?: any,
//   ref?: RefCallback,
//   children?: ThreshWidget[],
// }
export default class Widget <P = {}, S = {}> {
  props: P
  state?: S
  __vNode__: VNode

  static defaultProps: any = {}

  widgetDidMount() {}
  widgetDidUpdate() {}
  widgetDidUnmount() {}

  constructor (props: P) {
    if (checkObjectValid(props)) {
      this.props = Util.merge(props) as any
    }
  }
  
  /**
   * 更新state
   * @param {any} state
   */
  setState (state?: S) {
    if (!this.__vNode__.hasMount || !checkObjectValid(state)) return
    if (state) Object.assign(this.state, state)
    scheduleUpdate(this.__vNode__)
  }
  /**
   * 强制更新
   */
  forceUpdate() {
    this.setState()
  }

  /**
   * 组件渲染方法
   */
  render (): VNode | void {}
}

/**
 * 基础组件类
 * 继承自Widget
 * 外部不可使用该类
 */
export class BasicWidget <P = {}, S = {}> extends Widget <P, S> {
  static isBasicWidget: boolean = true
  static childrenRule: ChildrenRule

  setState () {
    threshApp.onError(new Error(`basic widget cannot call setState()`))
  }
}
