// thresh-lib/src/manager/UtilManager.ts
import MethodChannel, { FlutterMethodChannelType } from '../channel/MethodChannel'
import BridgeManager from './BridgeManager'
import Util from '../shared/Util'
import { RequestParams } from '../types/type'
import devtools from './DevtoolsManager'

/**
 * 工具方法管理器
 */
export default class UtilManager {
  /**
   * 发起网络请求
   * debugMode状态下会通过flutter发起请求
   * 否则通过native bridge发起请求
   */
  static async request (params: RequestParams, module: string = 'base', method: string = 'request'): Promise<any> {
    if (process.env.NODE_ENV !== 'production') {
      if (!BridgeManager.networkModuleNames.includes(module)) BridgeManager.networkModuleNames.push(module)
      if (!BridgeManager.networkModuleNames.includes(method)) BridgeManager.networkModuleNames.push(method)
    }
    return BridgeManager.invoke({
      module,
      method,
      params: params
    })
  }
  /**
   * 复制到剪贴板
   * @param {any} data
   */
  static copy (data: any, showSuccess: boolean = true) {
    MethodChannel.call({
      method: FlutterMethodChannelType.copy,
      params: {
        data: Util.toString(data),
        showSuccess
      },
    })
  }
  /**
   * 收起键盘
   */
  static blur () {
    MethodChannel.call({
      method: FlutterMethodChannelType.blur
    })
  }

  static log (...args: any[]) {
    if (console && console.log) {
      console.log(...args)
    }
    devtools.log(...args)
  }
  static warn (...args: any[]) {
    if (console && console.warn) {
      console.warn(...args)
    }
    devtools.warn(...args)
  }
  static error (...args: any[]) {
    if (console && console.error) {
      console.error(...args)
    }
    devtools.error(...args)
  }
}
